Fix GCE Cloud/Attach/Detach stability issues

saad-ali opened the pull request:
- Fixes #15978: GCE Cloudprovider `waitForOp` sometimes causes panics due to `invalid memory address or nil pointer dereference`
  - Fixes a bug where anytime there is a GCE error doing a get poll operation call the code would panic due to `nil pointer dereference`. This affects all GCE calls.
- Fixes #15382: Pod Disks should schedule a pod w/two RW PDs flake
  - Instead of using operation manager to defer to pending detach operations, this change ensures that attach/detach operations for the same GCE PD are atomic. This prevents a whole host of nasty race conditions.

This must be cherry picked for v1.1.
CC @thockin @justinsb

saad-ali committed:
Fix GCE Cloudprovider waitForOp bug and make GCE attach/detach atomic operations.

alex-mohr commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -54,6 +54,9 @@ const (
 	gceAffinityTypeClientIPProto = "CLIENT_IP_PROTO"
+
+	operationPollInterval        = 3 * time.Second
+	operationPollTimeoutDuration = 7 * time.Minute
```
where does the 7 minutes come from?  what's the behavior if 7 minutes is reached and e.g. operation actually succeeds (or fails) after 8 minutes?

saad-ali replied:
7 minutes is an arbitrary cap that I set, based on the fact that I've seen some disk attach request take a few minutes to succeed. I'm fine with either lowering or raising it. But, I think the poll time must be capped: if there is a server side issue, for example, where the operation is never set to "DONE" we don't want this to block indefinitely (as the old code would do).

When the timeout is hit, the request fails with error, and it is up to the caller to determine what to do (GCE PD attach/detach will retry).

alex-mohr replied:
It's generally a violation of the GCE api contract for an operation to never go to DONE, but there are known edge cases where it can take 15-20 minutes or even more to get there.

Does the pd retry code handle the case where there's an already-pending operation to detach a disk and GCE returns an error creating a new operation to do the same thing?  If not, then retries may not be improving reliability.

saad-ali replied:
> It's generally a violation of the GCE api contract for an operation to never go to DONE, but there are known edge cases where it can take 15-20 minutes or even more to get there.

I'm a big believer in defensive programming. Having a loop that can block indefinitely if the server returns the wrong thing or if there is a bug in the deserializer seems like a bad idea. I'm completely open to upping the limit to something very high... 30 minutes? Would you be ok with that?

> Does the pd retry code handle the case where there's an already-pending operation to detach a disk and GCE returns an error creating a new operation to do the same thing? If not, then retries may not be improving reliability.

The pd retry code would treat an already-pending operation as any other failure and would retry after some time. You're right that retries wouldn't improve reliability in that case. But they will eventually put the pod in a bad state, and generate events indicating that the pod is blocked because it is unable to attach the disk, because the attach timedout. Meaning the user eventually gets indication of what is going on, vs if this code blocked indefinitely the pod would be stuck forever in pending state, with no indication to the user what is happening.

alex-mohr replied:
Seems reasonable, thanks!  Given reasonable retry logic, 15 minutes seems fine.

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -245,45 +248,62 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
+	var pollOp *compute.Operation
+	var err error
+	return wait.Poll(operationPollInterval, operationPollTimeoutDuration, func() (bool, error) {
+		pollOp, err = getOperation()
```
Might be nice to at least log this error.

saad-ali replied:
Added logging inside `getErrorFromCompletedOp`

saad-ali replied:
And logged this swallowed error.

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -245,45 +248,62 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
+	}
+
+	var pollOp *compute.Operation
+	var err error
```
Is there any reason that these need to be declared outside and then closed over? Is it just a GC efficiency thing?

saad-ali replied:
Nope, just a leftover from when error handling was being done outside. Changed it.

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1442,18 +1462,7 @@ func (gce *GCECloud) AttachDisk(diskName string, readOnly bool) error {
-				return nil
-			}
-		}
+		return err
```
Was removing this intentional?

cjcullen replied:
I see now it probably was.

saad-ali replied:
Yep, intentional. If the disk is already attached, I want this to error out, and let the PD attach/detach code resolve it.

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -245,45 +248,62 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
 }
 
 func waitForOp(op *compute.Operation, getOperation func() (*compute.Operation, error)) error {
```
Could we just have `waitForOp` not take an operation? Or have `getOperation` be a `func(*compute.Operation) (*compute.Operation, error)`? Then you wouldn't have to check for nil in both places.

saad-ali replied:
Good idea, cleaned up.

saad-ali commented:
Thanks for the feedback @cjcullen and @alex-mohr! PTAL

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -244,47 +247,53 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
 func (gce *GCECloud) waitForGlobalOp(op *compute.Operation) error {
-	return waitForOp(op, func() (*compute.Operation, error) {
+	return waitForOp(op, func(operationName string) (*compute.Operation, error) {
 		return gce.service.GlobalOperations.Get(gce.projectID, op.Name).Do()
```
This will now crash here if operation is nil. If you make the func `func(op *compute.Operation) (*compute.Operation, error)`, then waitForOp can be the only one that needs to check for nil. Then you'd also need to handle errors a little bit differently above (you would only want to assign the return value of `getOperation` to `pollOp` if `err` was nil).

Or you could do it the opposite way: Check for nil in each of the three `waitFor***Op()` functions. Then have waitForOp just be:
```
func waitForOp(getOperation func() (*compute.Operation, error)) error {
	op, err := getOperation()
	...
}
```

saad-ali replied:
Whoops, I actually forgot to update `op.Name` to the new parameter `operationName`. The idea being that `waitFor...Op()` doesn't evaluate `op` therefore it doesn't need to do a nil check. Only `waitForOp()` does and it passes the `operationName` to the `getOperation()` function.

PTAL

saad-ali commented:
@justinsb Could you review the volume changes please?

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -244,48 +247,55 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
+	opName := op.Name
+	return wait.Poll(operationPollInterval, operationPollTimeoutDuration, func() (bool, error) {
+		pollOp, err := getOperation(opName)
+		glog.Warningf("GCE poll operation failed: %v", err)
```
```
if err != nil {
	glog.Warningf("GCE poll operation failed: %v", err)
}
```

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -244,48 +247,55 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
+	return op != nil && op.Status == "DONE"
+}
+
+func getErrorFromCompletedOp(op *compute.Operation) error {
```
This function is also used for non-completed ops. I'd call it getErrorFromOp, and leave a comment that the function will return nil if the op is nil or the op has no errors.

cjcullen commented:
LGTM on the GCE cloudprovider stuff after my last 2 nits.

saad-ali commented:
Thanks @cjcullen! I addressed the feedback and squashed the changes.

@justinsb please review the GCE volume changes when you get a chance. I'd like to get this in ASAP, since it is a 1.1 blocker. I've run overnight soak test and the changes make GCE volume attach/detach much more stable. Thanks

brendandburns commented:
LGTM.

alex-mohr commented:
Thanks for driving this Saad!

saad-ali commented:
@alex-mohr No problem. Could you confirm if you ok with: `operationPollTimeoutDuration = 10 * time.Minute`? Or should we increase it to 30 minutes?

alex-mohr commented:
@saad-ali Re: 10 minutes vs 30, if you have high confidence in the retry code handling simultaneous disk attach/detach operations correctly and that makes event handling better, 10 minutes should also work.  It's a persistent object in the GCE database that has a bunch of internal logic to drive it to completion (success or failure), so having multiple in flight at a time is unlikely to actually help and could even trigger less-tested code paths.  Your call, I'm fine with either.

saad-ali commented:
I changed it to from 10 minutes to 30 minutes to be safe--since there are some operations that are known to take a very long time.

PTAL and add a re-add the LGTM label.

lavalamp commented on pkg/util/keymutex/keymutex.go:
```diff
@@ -0,0 +1,82 @@
+
+type keyMutex struct {
+	sync.RWMutex
+	mutexMap map[string]*sync.Mutex
```
what prevents this from growing forever?

saad-ali replied:
Nothing. The map will track all the unique PDs seen by the kubelet during its runtime. However, I don't foresee this as being an issue unless someone is attaching/detaching millions of unique PDs (over the course of a a kubelet lifetime, which I'd guess is about a month for machine due to reboots required for patching), at which point they are likely to run in to GCE quota issues first.

alex-mohr replied:
FWIW, another time we decided that resource garbage collection wasn't important, it ended up biting us due to causing kubelet CPU usage to increase by 10% a day.

And if we make multiple such leak-100-bytes-per-pod-attempt decisions in various places, it'll become unworkable pretty quickly.

Is there a not-too-unreasonable fix for the leak?  File an issue or at least document the known-leaking behavior in comments?

saad-ali replied:
Opened up #16305 to track this.

lavalamp commented:
@k8s-bot test this please

lavalamp commented:
@k8s-bot unit test this please
